### ISSUE TITLE
Make eq a separate dimension

### ISSUE THREAD

**Issue by rossberg** [WebAssembly/gc#267]
Currently, `eq` is its own heap type, placed below `any` in the subtype hierarchy. That implies that any given form of reference type can only be comparable either always (a subtype of eqref) or never (not a subtype of eqref).

It is desirable to be able to distinguish comparable and incomparable references, for example:

* eq or non-eq function references (consider e.g. raw func refs vs closures),
* eq or non-eq struct references (consider structs representing object vs value types),
* eq or non-eq host references (consider implementation choices the host may or may not want to expose).

This can be solved by treating eq-ness as a dimension analogous to nullability. That is, the general form of reference type is
```
(ref null? eq? <heaptype>)
```
In exchange, the current heaptype `eq` is removed, and `eqref` becomes a shorthand for `(ref null eq any)`.

Note that this scheme could be extended further in the future, making the set of constraints on a reference a bitset. If designed properly, subtyping corresponds to simple bitset inclusion.

**Comment by askeksa-google** [comment]
There's a fundamental difference between `eq` on one hand and `null` and `i31` on the other.

Both `null` and `i31` each signify a set of values, and the presence of the type dimension indicates a (disjoint) union with that set.

From the description here, `ref eq T` does not mean "`T` or `eq`". Rather, `eq` is a property of a particular type (its values can be compared by reference).

Given that some values can be compared by reference and some can't, I have two questions for the proposal:
1. What does `ref eq T` mean if the values of `T` can't be compared by reference?
2. What is the utility of `ref non-eq U` if the values of `U` can be compared by reference? It only takes capabilities away from the type, AFAICS.

**Comment by jakobkummerow** [comment]
I'm not sure I find this idea convincing. As the motivational examples point out, comparability is a property of the _type_, not a property of _references_ to it. A given type `T` may be comparable or incomparable, but either way it doesn't seem to make sense to have both `(ref eq T)` and (incomparable) `(ref T)` in programs dealing with objects of type `T`.

The existing proposal with its `eqref` type at a certain point in the hierarchy kind of expressed one particular coarse/simplified assumption about which `T` are or aren't reference-comparable. If we think that that's too coarse, maybe the eq-bit should move to type definitions, i.e. `(type $Foo eq (struct (field ...)))`. However, if we do that, then the only way to express "any comparable reference" is for modules to define a custom otherwise-empty `(type $Comparable eq)` which they can use as supertypes for comparable struct/array/function types, at which point we might as well leave it around as a built-in type.

Another concern is that the `null` and `eq` attributes kind of work in opposite directions: for nullability, the more generic default is that a reference is nullable, and the "opt-in" special case is that a given reference is known not to be null. For comparability, however, the safe default is to assume incomparability, and the "opt-in" special case is that a given reference is known to be comparable. Put differently: the `eq` attribute is a capability of sorts, and in particular implicit upcasting from `(ref T)` to `(ref eq T)` is nonsensical (and potentially "dangerous" as far as introducing unintended behavior is concerned).

If a module _can_ freely use `(ref eq T)` for types `T` that aren't meaningfully reference-comparable, and avoiding this is left to the module's own carefulness, then that also puts a question mark behind the usefulness of the annotation in general. We might as well specify `ref.eq` to operate on `anyref`, and again leave it up to modules' own carefulness not to compare references that aren't meaningfully comparable.

**Comment by jakobkummerow** [comment]
Come to think of it, a slight tweak of the second paragraph of my post above might be a viable counter-proposal:

Since we have explicit supertypes now anyway, we could require custom type definitions to explicitly state whether they want to derive from `eqref` or not.

I'm not sure whether that leads to problems down the road; if you see any please don't hesitate to point them out.

Alternative counter-proposal: get rid of `eqref`, let `ref.eq` operate on any references, and explicitly specify canonicalization of funcrefs, i.e. `(ref.eq (ref.func $F) (ref.func $G)) == (i32.eq $F $G)`.

**Comment by rossberg** [comment]
@askeksa-google:
> There's a fundamental difference between `eq` on one hand and `null` and `i31` on the other.
> Both `null` and `i31` each signify a set of values, and the presence of the type dimension indicates a (disjoint) union with that set.

True, this is a generalisation of the notion of reference attribute.

> 1. What does `ref eq T` mean if the values of `T` can't be compared by reference?
> 2. What is the utility of `ref non-eq U` if the values of `U` can be compared by reference? It only takes capabilities away from the type, AFAICS.

I suspect you are conflating a reference value with what is referenced. Those are different categories. Whether a reference (aka pointer) is comparable is independent from what it's pointing to – in principle, all pointers could be compared. However, if some instruction or API creating a references does not want them to ever be comparable (i.e., to enable certain internal optimisations that pointer equality would leak), it would simply return only non-eq references to the underlying type.

@jakobkummerow:
> As the motivational examples point out, comparability is a property of the _type_, not a property of _references_ to it.

Quite the contrary: they are meant to point out that the same given type representation might be both used in a context where you want to expose equality on pointers to it or where you don't.

You could tie it to the type definition, and that indeed was my first idea. But I see no advantage in doing so. Making it orthogonal is more composable and also more natural – because the comparison is on pointers, not pointees.

> Another concern is that the `null` and `eq` attributes kind of work in opposite directions

You're right, the attribute and its textual rendering should probably be inverted to noeq. Alternatively, we could have `nonull`. Either way, the intention of course is that `eq <: noeq`.

> We might as well specify `ref.eq` to operate on `anyref`, and again leave it up to modules' own carefulness not to compare references that aren't meaningfully comparable.

That's an abstraction leak, it forces you into choosing between unspecified behaviour or preventing possible optimisations. For the language, itself the former is a no-go, and the latter undesirable. There has been agreement on the CG that we do not want to allow equality on (all?) function references. The ability to optimise value types without identity is a similar use case.

**Comment by askeksa-google** [comment]
This seems to come partly down to the philosophy about the purpose of Wasm. It's certainly my understanding in general that abstraction and encapsulation is meant to be handled at the source language level, while the responsibilities of Wasm are to be safe, well-defined and efficient. Under that assumption:

It is always safe to compare references. Since comparison of function references is (at the moment) unspecified, they are statically uncomparable in order to make Wasm well-defined. If a reference could be declared comparable independently of what it is pointing to, Wasm would no longer be well-defined.

Structs are comparable because their comparison is well-defined. A source-language compiler is free to produce code that depends on certain references not being compared, and it is its responsibility not to compare them. But because their comparison is well-defined at the Wasm level, it is useless (contributes to neither safeness, well-definedness nor efficiency) to declare them uncomparable at the Wasm level.

In principle, an engine could choose, when asked to allocate a non-comparable object containing only immutable fields, to re-use an existing object with the same fields values. This does seem like a niche use case, though, and it is not obvious that the underlying memoization could be done significantly more efficiently in the engine than in Wasm.

**Comment by rossberg** [comment]
> This seems to come partly down to the philosophy about the purpose of Wasm. It's certainly my understanding in general that abstraction and encapsulation is meant to be handled at the source language level, while the responsibilities of Wasm are to be safe, well-defined and efficient.

Yes, but in this case it also affects abstractions defined by the language itself, as the examples show.

> It is always safe to compare references. Since comparison of function references is (at the moment) unspecified, they are statically uncomparable in order to make Wasm well-defined. If a reference could be declared comparable independently of what it is pointing to, Wasm would no longer be well-defined.

Not so. If funcref is short for noeq, and ref.func continues to produce a noeq ref, then nothing changes, since you can never turn any of them into something comparable. But we have added the ability to introduce other producers for func refs later that create comparable ones (perhaps, e.g., func.bind).

> Structs are comparable because their comparison is well-defined. A source-language compiler is free to produce code that depends on certain references not being compared, and it is its responsibility not to compare them. But because their comparison is well-defined at the Wasm level, it is useless (contributes to neither safeness, well-definedness nor efficiency) to declare them uncomparable at the Wasm level.

Not at all! The relevant (and common) optimisation for value types (and why some languages have them) is that an engine can often eliminate allocations on the heap altogether. For example, a noeq ref to a small struct can be passed by copy, e.g., in registers. That's possible only if the struct has no observable identity.

We already discussed one particular case in the past: a (ref noeq (struct i32)) is essentially an i32ref and could be unboxed everywhere on 64 bit hardware. The eq version could not.

**Comment by titzer** [comment]
I'm not sure that we should add a whole dimension to the type system only to allow a *potential* future optimizations for non-comparable references. It seems like a complication that will just frustrate programmers/producers with irrelevant details most of the time, kind of like the "tail callable" attribute for functions that we rejected.

I think we should keep function references as non-comparable and leave the status quo. That's enough leeway for future optimizations IMHO. If you really want to compare functions in Wasm, you'd have to import a function to do it from the embedder. 

Keep in mind that in some embeddings, like JS, functions already do have identity, so I imagine that Web reality will force those engines to forgo optimizing `func.bind`. What would we do at that boundary for incomparable functions? I think the engine would be right back to square one.

**Comment by fgmccabe** [comment]
It is probably too late for the following suggestion; but I will make it anyway.
In languages like Java (and Haskell!), features like equality can be modeled by interfaces.

**Comment by tlively** [comment]
> > This seems to come partly down to the philosophy about the purpose of Wasm. It's certainly my understanding in general that abstraction and encapsulation is meant to be handled at the source language level, while the responsibilities of Wasm are to be safe, well-defined and efficient.
> 
> Yes, but in this case it also affects abstractions defined by the language itself, as the examples show.

@rossberg, could you elaborate on this? I agree with @askeksa-google's sentiment that this distinction should be handled by producers and could be erased in Wasm. I don't understand what you mean by affecting abstractions defined by the language.

I find the performance and optimization arguments more compelling so far, especially the possibility of passing and returning `noeq` structs in registers, but on the other hand that's a transformation that could be done by the producer rather than by the engine.

**Comment by titzer** [comment]
@tlively Yes, I agree that the optimizations around `noeq` structs can/should be mostly done by producers. For example, scalar replacement, hoisting allocations out of loops, and hash-consing I think are good candidates to do a layer up.

**Comment by rossberg** [comment]
Of course I agree that producers should perform all optimisations possible themselves. But some low-level optimisations depend on hardware. Since producers do not know their target hardware, there are some they cannot do. The unboxed i32ref example is a representative of one class of such examples that depends on the physical word size (and can be generalised to other small tuples). Others may depend on the number of registers available etc.

Things like (deterministic) pointer equality tend to get in the way of these optimisations, so it helps if producers can indicate whether they need it.

@tlively:
> @rossberg, could you elaborate on this? I agree with @askeksa-google's sentiment that this distinction should be handled by producers and could be erased in Wasm. I don't understand what you mean by affecting abstractions defined by the language.

If the abstraction in question is not one created by a producer but one of the language itself (like function references and certain engine optimisations on them), then it needs to be protected by the language itself.

> If you really want to compare functions in Wasm, you'd have to import a function to do it from the embedder.

As an aside, it's a bit frustrating that the habit of suggesting "solutions" along these lines still is so common. ;) There are Wasm embeddings today where producers have no way of introducing custom host functions, like they can in some web/JS environment. So, this is not a solution, and hasn't been for a long time. We ought to strive to make Wasm more self-contained, not less.

**Comment by jakobkummerow** [comment]
Re **optimization/performance benefits**:

V8 at least has no interest in automatic unboxing of small structs for the foreseeable future: we have 32-bit "compressed" pointers, and we like the memory savings benefits they bring much more than the extra encoding flexibility of 64-bit pointers. Maybe that'll change in a couple of years or decades, maybe it won't... I wouldn't hold my breath.

Re **`eq <: noeq` subtyping**:

I'm having a bit of trouble wrapping my head around it, but so far I don't see that as being useful (or even viable, depending on details). The `eq` bit on a reference is both a capability ("can perform fast pointer comparisons", fine to implicitly upcast away) and a restriction ("must not break identity by creating copies", _not_ fine to implicitly upcast away). _If_ there is a way to get back from `noeq` to `eq` (maybe a `ref.as_eq`, like we have `ref.as_non_null`?), then doing a round-trip through `noeq` typed code (which might well create copies) would break semantics of `eq`-comparable objects. So maybe there _shouldn't_ be a way to cast from `(ref noeq T)` to `(ref eq T)`, but in that case, wouldn't that nullify most benefits of being able to write generic code in terms of the supertype, and hence _effectively_ boil down to `(ref eq T)` and `(ref noeq T)` not having any subtyping relationship in either direction?

Re **funcref vs comparability**:

My suggestion to define funcrefs as comparable with built-in canonicalization (which, to emphasize, is the _opposite_ of the scenario that the CG with good reason decided against, namely JavaScript-style distinct-object identity of funcrefs) was due to the Binaryen team expressing frustration that this isn't possible today. Various patterns around calls (e.g. vtable implementations, speculative inlining/direct-calling with fallback, ...) could be optimized entirely by modules (i.e. without engine involvement) if only there was a way to compare that a given funcref references the same function as another funcref. The current "workaround" is to have a massive function table and, instead of working with funcrefs, identify functions by their index in this table (and then, accordingly, `call_indirect` instead of `call_ref`).

**Comment by rossberg** [comment]
> Re **`eq <: noeq` subtyping**:
> I'm having a bit of trouble wrapping my head around it, but so far I don't see that as being useful (or even viable, depending on details). The `eq` bit on a reference is both a capability ("can perform fast pointer comparisons", fine to implicitly upcast away) and a restriction ("must not break identity by creating copies", _not_ fine to implicitly upcast away). _If_ there is a way to get back from `noeq` to `eq` (maybe a `ref.as_eq`, like we have `ref.as_non_null`?), then doing a round-trip through `noeq` typed code (which might well create copies) would break semantics of `eq`-comparable objects.

Ah, I should have clarified that: of course there must not be a way to recover eq via down cast. I agree that would not make sense. As you say, it's a form of (static) capability. And as usual with true capabilities, if you need them, you have to keep/pass them around (statically, in this case), you cannot manufacture them retroactively.

> Re **funcref vs comparability**:
> My suggestion to define funcrefs as comparable with built-in canonicalization (which, to emphasize, is the _opposite_ of the scenario that the CG with good reason decided against, namely JavaScript-style distinct-object identity of funcrefs) was due to the Binaryen team expressing frustration that this isn't possible today.

Well, yes, that's one reason why it would be useful to be able to have both. Then a producer can pay as they go: if they request eq for a ref, then the engine has to be careful to maintain its semantics, potentially at a cost. If the producer does not care, then it can give the engine more leeway and save cost. Does that not give the best of both worlds?

**Comment by titzer** [comment]
If `eq` references are subtypes of `noeq` references, I'd be concerned not having a casting mechanism to recover that static information. The abstract reference types, e.g. `data`, do have instructions to dynamically recover lost static information, and it seems like a loss of symmetry to not have it. A casting mechanism would imply that objects would have dynamic type information to do this, maybe via the RTT?

**Comment by tlively** [comment]
@rossberg, would Wob or WAML benefit from making `eq` a separate dimension? In the abstract, I can see that some code may care about eq-ness and other code may not, but I'm having trouble coming up with an example where a producer would want to take advantage of that. And even if a producer would use this feature, it still would not be useful (except for expressing user-level safety properties, which we've generally taken to be a non-goal) if no engine uses it for optimizations.

**Comment by rossberg** [comment]
@titzer:
> If eq references are subtypes of noeq references, I'd be concerned not having a casting mechanism to recover that static information might be a mistake in the long run.

Well, in the design as given, that would defeat the purpose. But it is a good point. If we tied eqness to the heap type, then it could be considered in casts. That might be a good reason for not making it a dimension. I'll have to ponder about this a little.

@tlively: The compilers use single-field structs e.g. for boxing numbers and scalars with more than 31 bits. With this change, engines would be able to unbox those on 64-bit hardware (unless they use pointer compression), which currently they can't because they have to preserve reference identity. The most relevant case is int32. But that could also be achieved by specifying struct eqness by other means.
